### PR TITLE
Paths everywhere

### PR DESCRIPTION
**Pull request by cstamas**
Internally, stop using java.io.File and use NIO2 Paths instead. From outside, this should not make a lot of difference (and is enforced by japicmp as [documented here](https://maven.apache.org/resolver/api-compatibility.html)).

But, as a consequence: if Resolver used on FileSystem like Google JIMFS, whatever code touches File method (that will do Path.toFile() under the hood), it will explode. Hence, we need to make sure this call never happens internally, and leave it to consumer projects. Resolver alone should never invoke it.

Most notable changes:
* change internals to use NIO2 Paths.
* in resolver now everything is "resolved" from local repository `basePath`, hence, if you put your local repository onto non-default FileSystem, it will work just fine (as proved by Demo Snippets). Naturally, if anything touches `File` returning method, in case of non-default FileSystem, it explodes (Google JIMFs Path.toFile() simply throws).
* change transport-file to support any (and not only default) FileSystem, covered by tests. This makes possible to use any FileSystem hosted repositories as "remote file repositories". Covered by UTs that now runs as parameterized two times: using default and using JIMFS FileSystem (as it is simple to set up, very handy for testing).
* change file locking to support any (and not only default) FileSystem. Before this change the implicit requirement that "lock name is fully qualified absolute file system path", but now it changes to "lock name is proper string representation of an URI". The key change is in BasedirNameMapper, where one can now use custom `Path` (for example created with JIMFS) and use that to name locks (use `file:` for default FS or any other URIs, ie `jimfs:` ones).

---

If you build this locally, and then related [Maven PR](https://github.com/apache/maven/pull/1413) (~w/ japicmp disabled if not fixed~), and then uncomment Google FS in demo `Booter` class, you will end up with Resolver DEMOs running on in-memory FS.

---

https://issues.apache.org/jira/browse/MRESOLVER-496

### CODE REVIEW

**Comment by cstamas** [comment]
@gnodet @michael-o @slawekjaranowski both PRs (this and Maven https://github.com/apache/maven/pull/1413) are green, please eyeballs

**Review by gnodet — APPROVED** [review]
LGTM

File: maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultChecksumProcessor.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+    public DefaultChecksumProcessor(PathProcessor pathProcessor) {
+        this.pathProcessor = requireNonNull(pathProcessor);
+    }
+
+    @Override
+    public String readChecksum(final Path checksumPath) throws IOException {
+        // for now do exactly same as happened before, but FileProcessor is a component and can be replaced
+        String checksum = "";
+        try (BufferedReader br = new BufferedReader(
+                new InputStreamReader(Files.newInputStream(checksumPath), StandardCharsets.UTF_8), 512)) {
```
Comment (michael-o) [inline maven-resolver-impl/src/main/java/org/eclipse/aether/internal/impl/DefaultChecksumProcessor.java:55]: There is a better way for this: https://docs.oracle.com/javase/8/docs/api/java/nio/file/Files.html#newBufferedReader-java.nio.file.Path-java.nio.charset.Charset-

**Review by michael-o — COMMENTED** [review]
In future, when you request someone to review such a huge PR give people enough time to do it otherwise it does not make sense.

**Comment by cstamas** [comment]
> In future, when you request someone to review such a huge PR give people enough time to do it otherwise it does not make sense.

Sure, and sorry for that. I assumed a week is enough. 

OTOH, the longer the PR is held back, the new PRs are accumulating and as may saw. requires quite effort to sort out conflicts... So there are two sides of the coin.

**Comment by jira-importer** [comment]
Resolve #1170

**Pull request merged**